File: 332.reconstruct-itinerary.py
#
# @lc app=leetcode id=332 lang=python3
#
# [332] Reconstruct Itinerary
#

# @lc code=start
from collections import defaultdict, deque


class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        edges = defaultdict(list)
        for u, v in tickets:
            edges[u].append(v)
        for u in edges:  # less comparisons than full sort
            edges[u].sort(reverse=True)
        route = []

        # def dfs(u):
        #     while edges[u]:
        #         dfs(edges[u].pop())
        #     route.append(u)
        # dfs("JFK")

        
        s = deque(["JFK"])
        while s:
            while edges[s[-1]]:
                s.append(edges[s[-1]].pop())
            route.append(s.pop())
        return route[::-1]


# @lc code=end

File: 332.reconstruct-itinerary.rs
/*
 * @lc app=leetcode id=332 lang=rust
 *
 * [332] Reconstruct Itinerary
 */

// @lc code=start
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap, VecDeque};


impl Solution {
    pub fn find_itinerary(tickets: Vec<Vec<String>>) -> Vec<String> {
        let mut graph = HashMap::new();
    
        for (u, v) in tickets.iter().map(|v| (v[0].to_owned(), v[1].to_owned())) {
            graph.entry(u).or_insert(BinaryHeap::new()).push(Reverse(v));
        }
    
        fn dfs(
            graph: &mut HashMap<String, BinaryHeap<Reverse<String>>>,
            node: &str,
            route: &mut VecDeque<String>,
        ) {
            while let Some(next) = graph.get_mut(node).and_then(|dests| dests.pop()) {
                dfs(graph, &next.0, route);
            }
            route.push_front(node.to_string());
        }
    
        let mut route = VecDeque::new();
        dfs(&mut graph, "JFK", &mut route);
        route.into()
    }
}
// @lc code=end
 